src/main/java/build/buildfarm/common/resources/ResourceParser.java

```diff
@@ -308,6 +309,11 @@ private static BlobInformation parseBlobInformation(String[] segments, MutableIn
     if (DIGEST_FUNCTIONS.containsKey(maybeDigestFunction)) {
       builder.setDigestFunction(DIGEST_FUNCTIONS.get(maybeDigestFunction));
       index.getAndIncrement();
+    } else {
+      // Infer the digest function based on the hash format
+      // Should go here when it's not BLAKE3
+      String hash = maybeDigestFunction;
+      builder.setDigestFunction(DigestUtil.HashFunction.forHash(hash).getDigestFunction());
```

Comment:
DigestFunction is always included in the request, this additional handling might not be needed.
There are several other hashing algorithms with the same hash length as BLAKE3. It might be better to remove the HashFunction.forHash method entirely and always rely on the request.

Reply:
I don't see it always included in the request. For blake it is, for sha it isn't.
I assume if we want to change that we would need to change the Bazel OSS first

Reply:
I thought that hash function is not part of request only for SHA256 as SHA256 is default hashing algorithm for bazel. However, upon checking the Bazel code, I found that it classifies hashing algorithms into two groups: old and new. For the old ones, where the hashing algorithm can be identified using the hash length, the hash function is not part of the request. For the new ones, the hash function is included in the request. The hash detection logic should complement this Bazel method.
https://github.com/bazelbuild/bazel/blob/1dd5c20513b6ff479bf796fe2541bd492c661c94/src/main/java/com/google/devtools/build/lib/remote/ByteStreamUploader.java#L183

Reply:
It is complementing that now no? For new one we extract if it's part of the request, otherwise we determine by length with this function I added:
```java
public static HashFunction forHash(String hexDigest) {
      int hashLength = hexDigest.length();
        return switch (hashLength) {
            case 64 -> // 32 bytes * 2 characters per byte
                    HashFunction.SHA256;
            case 40 -> // 20 bytes * 2 characters per byte
                    HashFunction.SHA1;
            case 32 -> // 16 bytes * 2 characters per byte
                    HashFunction.MD5;
            default -> throw new IllegalArgumentException("hash type unrecognized: " + hexDigest);
        };
```